src/idiomatic/leveraging-the-type-system/raii/drop_limitations.md

```diff
@@ -0,0 +1,84 @@
+# The limitations of `Drop`
+
+While `Drop` works well for cases like synchronization primitives, its use
+becomes more questionable when dealing with I/O or unsafe resources.
+
+```rust
+use std::fs::File;
+use std::io::{self, Write};
+
+fn write_log() -> io::Result<()> {
+    let mut file = File::create("log.txt")?;
+    // ^ ownership of the (OS) file handle starts here
+
+    writeln!(file, "Logging a message...")?;
+    Ok(())
+} // file handle goes out of scope here
+```
+
+<details>
+
+- In the earlier example, our `File` resource owns a file handle provided by the
+  operating system.
+
+  [As stated in the documentation](https://doc.rust-lang.org/std/fs/struct.File.html):
+
+  > Files are automatically closed when they go out of scope. Errors detected on
+  > closing are ignored by the implementation of Drop.
+
+- This highlights a key limitation of the `Drop` trait: it cannot propagate
+  errors to the caller. In other words, fallible cleanup logic cannot be handled
+  by the code using the `File`.
+
+  This becomes clear when looking at the
+  [definition of the `Drop` trait](https://doc.rust-lang.org/std/ops/trait.Drop.html):
+
+  ```rust
+  trait Drop {
+      fn drop(&mut self);
+  }
+  ```
+
+  Since `drop` does not return a `Result`, any error that occurs during cleanup
+  cannot be surfaced or recovered from. This is by design: `drop` is invoked
+  automatically when a value is popped off the stack during unwinding, leaving
+  no opportunity for error handling.
+
+- One workaround is to panic inside `drop` when a failure occurs. However, this
+  is risky—if a panic happens while the stack is already unwinding, the program
+  will abort immediately, and remaining resources will not be cleaned up.
+
+  While panicking in `drop` can serve certain purposes (see
+  [the next chapter on "drop bombs"](./drop_bomb.md)), it should be used
+  sparingly and with full awareness of the consequences.
+
+- Another drawback of `drop` is that its execution is implicit and
+  non-deterministic in terms of timing. You cannot control _when_ a value is
+  dropped. And in fact as discussed in previous slide it might never even run at
+  all, leaving the external resource in an undefined state.
+
+  This matters particularly for I/O: normally you might set a timeout on
+  blocking operations, but when I/O occurs in a `drop` implementation, you have
+  no way to enforce such constraints.
+
+  Returning to the `File` example: if the file handle hangs during close (e.g.,
+  due to OS-level buffering or locking), the drop operation could block
+  indefinitely. Since the call to `drop` happens implicitly and outside your
+  control, there's no way to apply a timeout or fallback mechanism.
+
+- For smart pointers and synchronization primitives, none of these drawbacks
+  matter, since the operations are nearly instant and a program panic does not
+  cause undefined behavior. The poisoned state disappears along with the
```

Comment:
Note that the chapter does not discuss poisoned mutexes at the moment (I'm requesting that to be added in my comments above).